File: tools/fbx_stats.py
"""Budget stats for a binary FBX (stdlib only; no FBX SDK or Blender needed).

Usage: python tools/fbx_stats.py file.fbx [more.fbx ...] [--json]

Reports per mesh: triangles (polygons fan-triangulated), distinct materials,
blendshape targets; and totals for meshes, triangles, material slots (≈ draw
calls), unique materials, bones (LimbNode), blendshapes, textures. Numbers are
what design §4 budgets are checked against.
"""

from __future__ import annotations

import json
import struct
import sys
import zlib
from typing import Any

_ARRAY_FMT = {"f": "f", "d": "d", "l": "q", "i": "i", "b": "?"}


class Node:
    __slots__ = ("name", "props", "children")

    def __init__(self, name: str, props: list, children: list):
        self.name = name
        self.props = props
        self.children = children

    def child(self, name: str) -> "Node | None":
        return next((c for c in self.children if c.name == name), None)

    def all(self, name: str) -> list["Node"]:
        return [c for c in self.children if c.name == name]


def _read_props(buf: bytes, pos: int, count: int) -> tuple[list, int]:
    props: list = []
    for _ in range(count):
        t = chr(buf[pos]); pos += 1
        if t == "Y":
            props.append(struct.unpack_from("<h", buf, pos)[0]); pos += 2
        elif t == "C":
            props.append(bool(buf[pos])); pos += 1
        elif t == "I":
            props.append(struct.unpack_from("<i", buf, pos)[0]); pos += 4
        elif t == "F":
            props.append(struct.unpack_from("<f", buf, pos)[0]); pos += 4
        elif t == "D":
            props.append(struct.unpack_from("<d", buf, pos)[0]); pos += 8
        elif t == "L":
            props.append(struct.unpack_from("<q", buf, pos)[0]); pos += 8
        elif t in _ARRAY_FMT:
            n, enc, clen = struct.unpack_from("<III", buf, pos); pos += 12
            raw = buf[pos:pos + clen]; pos += clen
            if enc == 1:
                raw = zlib.decompress(raw)
            fmt = _ARRAY_FMT[t]
            props.append(struct.unpack(f"<{n}{fmt}", raw[: n * struct.calcsize(fmt)]))
        elif t in ("S", "R"):
            n = struct.unpack_from("<I", buf, pos)[0]; pos += 4
            data = buf[pos:pos + n]; pos += n
            props.append(data.decode("utf-8", "replace") if t == "S" else data)
        else:
            raise ValueError(f"Unknown FBX property type {t!r} at {pos - 1}")
    return props, pos


def _read_node(buf: bytes, pos: int, wide: bool) -> tuple[Node | None, int]:
    if wide:
        end, nprops, _plen = struct.unpack_from("<QQQ", buf, pos); pos += 24
    else:
        end, nprops, _plen = struct.unpack_from("<III", buf, pos); pos += 12
    name_len = buf[pos]; pos += 1
    if end == 0:
        return None, pos
    name = buf[pos:pos + name_len].decode("ascii", "replace"); pos += name_len
    props, pos = _read_props(buf, pos, nprops)
    children: list[Node] = []
    while pos < end:
        child, pos = _read_node(buf, pos, wide)
        if child is None:
            break
        children.append(child)
    return Node(name, props, children), end


def parse(path: str) -> list[Node]:
    buf = open(path, "rb").read()
    if not buf.startswith(b"Kaydara FBX Binary"):
        raise ValueError(f"{path}: not a binary FBX")
    version = struct.unpack_from("<I", buf, 23)[0]
    wide = version >= 7500
    pos, nodes = 27, []
    while pos < len(buf):
        node, pos = _read_node(buf, pos, wide)
        if node is None:
            break
        nodes.append(node)
    return nodes


def _obj_name(node: Node) -> str:
    return str(node.props[1]).split("\x00")[0] if len(node.props) > 1 else "?"


def stats(path: str) -> dict[str, Any]:
    top = {n.name: n for n in parse(path)}
    objects = top["Objects"].children
    conns = [c.props for c in top["Connections"].all("C")] if "Connections" in top else []

    by_id = {o.props[0]: o for o in objects if o.props}
    parent_of: dict[int, list[int]] = {}
    children_of: dict[int, list[int]] = {}
    for p in conns:
        if len(p) >= 3 and p[0] in ("OO", "OP"):
            parent_of.setdefault(p[1], []).append(p[2])
            children_of.setdefault(p[2], []).append(p[1])

    def kind(o: Node) -> str:
        return str(o.props[2]) if len(o.props) > 2 else ""

    models = {oid: o for oid, o in by_id.items() if o.name == "Model"}
    mesh_models = {oid: o for oid, o in models.items() if kind(o) == "Mesh"}
    bones = [o for o in models.values() if kind(o) == "LimbNode"]
    materials = {oid: o for oid, o in by_id.items() if o.name == "Material"}
    textures = [o for o in by_id.values() if o.name == "Texture"]

    meshes: list[dict[str, Any]] = []
    for mid, model in mesh_models.items():
        geoms = [by_id[c] for c in children_of.get(mid, []) if c in by_id and by_id[c].name == "Geometry"]
        mats = [c for c in children_of.get(mid, []) if c in materials]
        tris = 0
        shapes = 0
        for g in geoms:
            pvi = g.child("PolygonVertexIndex")
            if pvi is not None:
                n = 0
                for v in pvi.props[0]:
                    n += 1
                    if v < 0:
                        tris += max(n - 2, 0)
                        n = 0
            # Geometry <- BlendShape deformer <- BlendShapeChannel(s)
            for d in children_of.get(g.props[0], []):
                if d in by_id and kind(by_id[d]) == "BlendShape":
                    shapes += sum(1 for ch in children_of.get(d, []) if ch in by_id and kind(by_id[ch]) == "BlendShapeChannel")
        meshes.append({
            "mesh": _obj_name(model),
            "triangles": tris,
            "materials": [_obj_name(materials[m]) for m in mats],
            "blendshapes": shapes,
        })

    meshes.sort(key=lambda m: -m["triangles"])
    return {
        "file": path,
        "meshes": len(meshes),
        "triangles": sum(m["triangles"] for m in meshes),
        "material_slots": sum(len(m["materials"]) for m in meshes),
        "unique_materials": len({n for m in meshes for n in m["materials"]}),
        "bones": len(bones),
        "blendshapes": sum(m["blendshapes"] for m in meshes),
        "textures": len(textures),
        "per_mesh": meshes,
    }


def main(argv: list[str]) -> None:
    as_json = "--json" in argv
    files = [a for a in argv if a != "--json"]
    results = [stats(f) for f in files]
    if as_json:
        print(json.dumps(results, indent=2))
        return
    for r in results:
        print(f"{r['file']}\n  meshes {r['meshes']}  triangles {r['triangles']}  material slots {r['material_slots']}"
              f"  unique materials {r['unique_materials']}  bones {r['bones']}  blendshapes {r['blendshapes']}"
              f"  textures {r['textures']}")
        for m in r["per_mesh"]:
            print(f"    {m['mesh']:<28} tris {m['triangles']:>7}  mats {len(m['materials'])}  shapes {m['blendshapes']}")


if __name__ == "__main__":
    main(sys.argv[1:])

File: src/fbx-stats.ts
/**
 * Budget stats for a binary FBX, no FBX SDK needed (TypeScript port of tools/fbx_stats.py).
 *
 * Triangles are counted by fan-triangulating each polygon; material slots (one per
 * mesh × material) approximate Unity draw calls for a skinned character.
 */

import fs from "node:fs";
import zlib from "node:zlib";

interface FbxNode {
  name: string;
  props: unknown[];
  children: FbxNode[];
}

const ARRAY_TYPES: Record<string, [number, (b: Buffer, o: number) => number | bigint | boolean]> = {
  f: [4, (b, o) => b.readFloatLE(o)],
  d: [8, (b, o) => b.readDoubleLE(o)],
  l: [8, (b, o) => b.readBigInt64LE(o)],
  i: [4, (b, o) => b.readInt32LE(o)],
  b: [1, (b, o) => b[o] !== 0],
};

function readProps(buf: Buffer, pos: number, count: number): [unknown[], number] {
  const props: unknown[] = [];
  for (let n = 0; n < count; n++) {
    const t = String.fromCharCode(buf[pos]);
    pos += 1;
    switch (t) {
      case "Y": props.push(buf.readInt16LE(pos)); pos += 2; break;
      case "C": props.push(buf[pos] !== 0); pos += 1; break;
      case "I": props.push(buf.readInt32LE(pos)); pos += 4; break;
      case "F": props.push(buf.readFloatLE(pos)); pos += 4; break;
      case "D": props.push(buf.readDoubleLE(pos)); pos += 8; break;
      case "L": props.push(buf.readBigInt64LE(pos)); pos += 8; break;
      case "S":
      case "R": {
        const len = buf.readUInt32LE(pos);
        pos += 4;
        const data = buf.subarray(pos, pos + len);
        pos += len;
        props.push(t === "S" ? data.toString("utf8") : data);
        break;
      }
      default: {
        const spec = ARRAY_TYPES[t];
        if (!spec) throw new Error(`Unknown FBX property type '${t}' at ${pos - 1}`);
        const length = buf.readUInt32LE(pos);
        const encoding = buf.readUInt32LE(pos + 4);
        const clen = buf.readUInt32LE(pos + 8);
        pos += 12;
        let raw = buf.subarray(pos, pos + clen);
        pos += clen;
        if (encoding === 1) raw = zlib.inflateSync(raw);
        const [size, read] = spec;
        const arr = new Array(length);
        for (let i = 0; i < length; i++) arr[i] = read(raw, i * size);
        props.push(arr);
      }
    }
  }
  return [props, pos];
}

function readNode(buf: Buffer, pos: number, wide: boolean): [FbxNode | null, number] {
  let end: number, nprops: number;
  if (wide) {
    end = Number(buf.readBigUInt64LE(pos));
    nprops = Number(buf.readBigUInt64LE(pos + 8));
    pos += 24;
  } else {
    end = buf.readUInt32LE(pos);
    nprops = buf.readUInt32LE(pos + 4);
    pos += 12;
  }
  const nameLen = buf[pos];
  pos += 1;
  if (end === 0) return [null, pos];
  const name = buf.subarray(pos, pos + nameLen).toString("latin1");
  pos += nameLen;
  const [props, afterProps] = readProps(buf, pos, nprops);
  pos = afterProps;
  const children: FbxNode[] = [];
  while (pos < end) {
    const [child, next] = readNode(buf, pos, wide);
    pos = next;
    if (!child) break;
    children.push(child);
  }
  return [{ name, props, children }, end];
}

export function parseFbx(buf: Buffer): FbxNode[] {
  if (buf.subarray(0, 18).toString("latin1") !== "Kaydara FBX Binary") throw new Error("Not a binary FBX");
  const wide = buf.readUInt32LE(23) >= 7500;
  const nodes: FbxNode[] = [];
  let pos = 27;
  while (pos < buf.length) {
    const [node, next] = readNode(buf, pos, wide);
    pos = next;
    if (!node) break;
    nodes.push(node);
  }
  return nodes;
}

export interface MeshStats {
  mesh: string;
  triangles: number;
  materials: string[];
  blendshapes: number;
}

export interface FbxStats {
  meshes: number;
  triangles: number;
  material_slots: number;
  unique_materials: number;
  bones: number;
  blendshapes: number;
  textures: number;
  per_mesh: MeshStats[];
}

const objName = (n: FbxNode) => String(n.props[1] ?? "?").split("\u0000")[0];
const kind = (n: FbxNode) => String(n.props[2] ?? "");

export function fbxStats(buf: Buffer): FbxStats {
  const top = new Map(parseFbx(buf).map((n) => [n.name, n]));
  const objects = top.get("Objects")?.children ?? [];
  const conns = (top.get("Connections")?.children ?? []).filter((c) => c.name === "C").map((c) => c.props);
  const byId = new Map<bigint, FbxNode>();
  for (const o of objects) if (typeof o.props[0] === "bigint") byId.set(o.props[0] as bigint, o);
  const childrenOf = new Map<bigint, bigint[]>();
  for (const p of conns) {
    if ((p[0] === "OO" || p[0] === "OP") && typeof p[1] === "bigint" && typeof p[2] === "bigint") {
      const list = childrenOf.get(p[2]) ?? [];
      list.push(p[1]);
      childrenOf.set(p[2], list);
    }
  }
  const models = [...byId.entries()].filter(([, o]) => o.name === "Model");
  const bones = models.filter(([, o]) => kind(o) === "LimbNode").length;
  const textures = objects.filter((o) => o.name === "Texture").length;

  const perMesh: MeshStats[] = [];
  for (const [mid, model] of models.filter(([, o]) => kind(o) === "Mesh")) {
    const kids = childrenOf.get(mid) ?? [];
    const geoms = kids.map((k) => byId.get(k)).filter((o): o is FbxNode => o?.name === "Geometry");
    const materials = kids.map((k) => byId.get(k)).filter((o): o is FbxNode => o?.name === "Material").map(objName);
    let triangles = 0;
    let blendshapes = 0;
    for (const g of geoms) {
      const pvi = g.children.find((c) => c.name === "PolygonVertexIndex");
      if (pvi) {
        let n = 0;
        for (const v of pvi.props[0] as number[]) {
          n++;
          if (v < 0) {
            triangles += Math.max(n - 2, 0);
            n = 0;
          }
        }
      }
      for (const d of childrenOf.get(g.props[0] as bigint) ?? []) {
        const deformer = byId.get(d);
        if (deformer && kind(deformer) === "BlendShape") {
          blendshapes += (childrenOf.get(d) ?? []).filter((c) => byId.get(c) && kind(byId.get(c)!) === "BlendShapeChannel").length;
        }
      }
    }
    perMesh.push({ mesh: objName(model), triangles, materials, blendshapes });
  }
  perMesh.sort((a, b) => b.triangles - a.triangles);
  return {
    meshes: perMesh.length,
    triangles: perMesh.reduce((s, m) => s + m.triangles, 0),
    material_slots: perMesh.reduce((s, m) => s + m.materials.length, 0),
    unique_materials: new Set(perMesh.flatMap((m) => m.materials)).size,
    bones,
    blendshapes: perMesh.reduce((s, m) => s + m.blendshapes, 0),
    textures,
    per_mesh: perMesh,
  };
}

export function fbxStatsFromFile(path: string): FbxStats {
  return fbxStats(fs.readFileSync(path));
}

/** Design §4 character budgets per LOD. */
export const LOD_BUDGETS = {
  LOD0: { triangles: 60_000, material_slots: 8, bones: null as number | null },
  LOD1: { triangles: 25_000, material_slots: 5, bones: 100 },
  LOD2: { triangles: 10_000, material_slots: 2, bones: null as number | null },
} as const;
export type LodLabel = keyof typeof LOD_BUDGETS;

export interface BudgetCheck {
  lod: LodLabel;
  checks: Array<{ metric: string; value: number; limit: number; pass: boolean }>;
  pass: boolean;
}

export function checkBudget(stats: FbxStats, lod: LodLabel): BudgetCheck {
  const b = LOD_BUDGETS[lod];
  const checks = [
    { metric: "triangles", value: stats.triangles, limit: b.triangles },
    { metric: "material_slots (≈ draw calls)", value: stats.material_slots, limit: b.material_slots },
    ...(b.bones !== null ? [{ metric: "bones", value: stats.bones, limit: b.bones }] : []),
  ].map((c) => ({ ...c, pass: c.value <= c.limit }));
  return { lod, checks, pass: checks.every((c) => c.pass) };
}
